### PR TITLE
[refactor][ir] Simplify ir.h to reduce compile-time

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #933

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

**Comment by archibate** [comment]
Build passing! @xumingkuan You were right in https://github.com/taichi-dev/taichi/issues/933#issuecomment-626098467!

File: taichi/ir/ir.h [hunk 1]
```diff
@@ -1552,5 +1552,4 @@ inline void StmtFieldManager::operator()(const char *key, T &&value) {
 
 TLANG_NAMESPACE_END
 
-#include "taichi/ir/statements.h"
 #include "taichi/ir/visitors.h"
```
Comment (xumingkuan) [inline taichi/ir/ir.h:1555]: What about also remove this? We need to include `visitors.h` in some but not all files that include `ir.h`.

**Comment by archibate** [comment]
Maybe we should move `Stmt` to `taichi/ir/stmt.h`?

**Comment by xumingkuan** [comment]
My benchmark:
Before this PR: build all 2m57s, touch ir.h 2m10s
After this PR: build all 2m48s, touch ir.h 1m51s

I think this is satisfying. Cool!

File: taichi/ir/analysis.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  bool certain() {
+    TI_ASSERT(related);
+    return high == low + 1;
+  }
+};
+
+// IR Analysis
+namespace irpass {
```
Comment (xumingkuan) [inline taichi/ir/analysis.h:51]: We can use `namespace irpass::analysis` here.

**Comment by xumingkuan** [comment]
Build time of touching `passes.h`: 53s
Build time of touching `analysis.h`: 36s

Great!

**Comment by xumingkuan** [comment]
> Do we need to separate `passes.h` and `analysis.h`? And if we do, maybe `transforms.h` is a better name for "passes"? `analysis.h` contains analysis passes, I think.

@yuanming-hu What do you think about the file structure here?

File: taichi/ir/analysis.h (hunk 2 above)
Comment (archibate) [reply taichi/ir/analysis.h:51]: Will this cause an error if `namespace irpass` not defined yet?

**Comment by archibate** [comment]
> Build time of touching `passes.h`: 53s
> Build time of touching `analysis.h`: 36s
> 
> Great!

Does this helpful for your work? Could you also test this on `master` and see if we make a difference?

**Comment by yuanming-hu** [comment]
> > Do we need to separate `passes.h` and `analysis.h`? And if we do, maybe `transforms.h` is a better name for "passes"? `analysis.h` contains analysis passes, I think.
> 
> @yuanming-hu What do you think about the file structure here?

I believe you guys can make a your own decision here. I'll champion whatever you guys end up with.

**Comment by archibate** [comment]
> Maybe we should move `Stmt` to `taichi/ir/stmt.h`?

It's hard to make things perfect, given the `ti format` status, let's first merge this and deal with stmt.h in a separate PR.

**Comment by xumingkuan** [comment]
> > Maybe we should move `Stmt` to `taichi/ir/stmt.h`?
> 
> It's hard to make things perfect, given the `ti format` status, let's first merge this and deal with stmt.h in a separate PR.


I think it's good to use the existing `taichi/ir/statements.h`. And let's merge this first.

**Comment by archibate** [comment]
So we will merge this once CI passed?

**Pull request merged**